From 5853b506443ce9d8a6bb0b477050d49f6c83385f Mon Sep 17 00:00:00 2001
From: Nan Wu <wnan@google.com>
Date: Mon, 3 Jun 2024 19:12:22 +0000
Subject: [PATCH] RESTRICT AUTOMERGE Allow an activity to start a new task via
 startNextMatchingActivity

A vulnerability bug fix (b/230492947) causes this no longer works.
We cannot simply fix the issue by trying to avoid move the task to
the front. The real reason is that startNextMatchingActivity uses
launchedFromUid (the app the lauches the orignal activity) to as
the callingUid and realCallingUid to launch the next matching
activity - though setting callingUid is largely irrelevant because
ActivityStarter sets it to caller app's uid if the caller is set;
In this case, the caller is set to the app that calls the method.
But for BAL purpose, this is not good because the launched
from app might have privileges, such as the Launcher app. Change the
realCallingUid to the activity that calls startNextMatchingActivity.

Bug: 337726734
Test: Manual test. BackgroundActivityLaunchTest#testBackgroundActivityBlockedInStartNextMatchingActivity
Flag: EXEMPT Bug Fix
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:d34ae40f870d4362a069940a035a4d58a536a231)
Merged-In: Ia8c6123e8e9325d17d2bfab9589b11eb087da97a
Change-Id: Ia8c6123e8e9325d17d2bfab9589b11eb087da97a
---
 .../server/wm/ActivityTaskManagerService.java | 41 +++++++++++--------
 1 file changed, 25 insertions(+), 16 deletions(-)

diff --git a/services/core/java/com/android/server/wm/ActivityTaskManagerService.java b/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
index d7489f2c7b91..fae7b4dc1a10 100644
--- a/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
+++ b/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
@@ -1411,6 +1411,8 @@ public class ActivityTaskManagerService extends IActivityTaskManager.Stub {
                 resultTo.removeResultsLocked(r, resultWho, requestCode);
             }
 
+            final int origCallingUid = Binder.getCallingUid();
+            final int origCallingPid = Binder.getCallingPid();
             final long origId = Binder.clearCallingIdentity();
             // TODO(b/64750076): Check if calling pid should really be -1.
             try {
@@ -1424,23 +1426,30 @@ public class ActivityTaskManagerService extends IActivityTaskManager.Stub {
                 // if the intent used matches both:
                 // - the activity in the background
                 // - a second activity.
-                options.getOptions(r).setAvoidMoveToFront();
+                // Fixes b/230492947 b/337726734
+                // Prevents background activity launch through #startNextMatchingActivity
+                // launchedFromUid of the calling activity represents the app that launches it.
+                // It may have BAL privileges (i.e. the Launcher App). Using its identity to
+                // launch to launch next matching activity causes BAL.
+                // Change the realCallingUid to the calling activity's uid.
+                // In ActivityStarter, when caller is set, the callingUid and callingPid are
+                // ignored. So now both callingUid and realCallingUid is set to the caller app.
                 final int res = getActivityStartController()
-                        .obtainStarter(intent, "startNextMatchingActivity")
-                        .setCaller(r.app.getThread())
-                        .setResolvedType(r.resolvedType)
-                        .setActivityInfo(aInfo)
-                        .setResultTo(resultTo != null ? resultTo.token : null)
-                        .setResultWho(resultWho)
-                        .setRequestCode(requestCode)
-                        .setCallingPid(-1)
-                        .setCallingUid(r.launchedFromUid)
-                        .setCallingPackage(r.launchedFromPackage)
-                        .setCallingFeatureId(r.launchedFromFeatureId)
-                        .setRealCallingPid(-1)
-                        .setRealCallingUid(r.launchedFromUid)
-                        .setActivityOptions(options)
-                        .execute();
+                            .obtainStarter(intent, "startNextMatchingActivity")
+                            .setCaller(r.app.getThread())
+                            .setResolvedType(r.resolvedType)
+                            .setActivityInfo(aInfo)
+                            .setResultTo(resultTo != null ? resultTo.token : null)
+                            .setResultWho(resultWho)
+                            .setRequestCode(requestCode)
+                            .setCallingPid(-1)
+                            .setCallingUid(r.launchedFromUid)
+                            .setCallingPackage(r.launchedFromPackage)
+                            .setCallingFeatureId(r.launchedFromFeatureId)
+                            .setRealCallingPid(origCallingPid)
+                            .setRealCallingUid(origCallingUid)
+                            .setActivityOptions(options)
+                            .execute();
                 r.finishing = wasFinishing;
                 return res == ActivityManager.START_SUCCESS;
             } finally {
-- 
2.34.1

